keps/sig-node/4680-add-resource-health-to-pod-status/README.md

```diff
@@ -0,0 +1,421 @@
+
+The proposal is to update the Pod Status with the device status even if the pod has been marked as Failed already, but still known to the kubelet.
+
+This use case is important to explore so this status may inform the retry policy introduced by the KEP: [Retriable and non-retriable Pod failures for Jobs](https://github.com/kubernetes/enhancements/issues/3329).
+
+
+***Do we need the CheckDeviceHealth call introduced to the Device Plugin to work around the limitation above?***
+
+We may consider this as a future improvement. 
+
+
+***Should we introduce a permanent failure status?***
+
+We may consider this as a future improvement. 
+
+### User Stories (Optional)
+
+#### Story 1
+
+- User scheduled a Pod using the GPU device
+- When GPU device fails, user sees the Pod is in crash loop backoff
+- User checks the Pod Status using `kubectl describe pod`
+- User sees the pod status indicating that the GPU device is not healthy
+- User or some (custom for now) controller deletes the Pod and replicaset reschedules it on another available GPU
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+There is not many risks of this KEP. The biggest risk is that Device Plugins will not be
+able to detect device health reliable and fast enough to assign this status to the
+Pods, marked as `restartPolicy=Never`. End users will expect this field and the
+missing health information will confuse them.
+
+## Design Details
+
+### Device Plugin implementation details
+
+Kubelet already keeps track of healthy and unhealthy devices as well as the mapping of those devices to Pods.
+
+One improvement will be needed is to distinguish unhealthy devices (marked unhealthy explicitly) and when device plugin was unregistered.
+
+NVIDIA device plugin has the checkHealth implementation: https://github.com/NVIDIA/k8s-device-plugin/blob/eb3a709b1dd82280d5acfb85e1e942024ddfcdc6/internal/rm/health.go#L39 that has more information than simple “Unhealthy”.
+
+We should consider introducing another field to the Status that will be a free form error information as a future improvement.
+### DRA implementation details
+
+Today DRA does not return the health of the device back to kubelet. The proposal is to extend the
+
+type `NamedResourcesInstance` (from [pkg/apis/resource/namedresources.go](https://github.com/kubernetes/kubernetes/blob/790dfdbe386e4a115f41d38058c127d2dd0e6f44/pkg/apis/resource/namedresources.go#L29-L37)) to include the Health field the same way it is done in 
```

Comment:
This is not part of our current KEP updates. You don't actually need this reported to the control plane for this KEP to work, right? You just need to be able to query the DRA driver for the info?

cc @pohly

Reply:
The device plugin API doesn't (technically) report health status either.

The "health" of a device is instead inferred as follows:
* A persistent `ListAndWatch()` call is made from the kubelet to a plugin, over which a stream of device IDs is passed and then tracked by the kubelet
* The initial call establishes a set of "healthy" devices that the plugin knows about and tracks
* At any point in time the plugin ay decide to send an updated stream of device IDs
* For each update, the list of already tracked device IDs is compared against the new list
* If a previously tracked device ID is missing -- it is marked as unhealthy
* If a new device ID is now present, the kubelet begins to track it (as healthy initially)

In 1.30 we had a similar `ListAndWatch()` API, from which we could have inferred something very analogous to the above. However, we are removing this in 1.31, so will need to provide something different.

Does it makes sense to keep some sort of stream like this, but make it _*just*_ about health status information? Meaning we stream back a list of deviceIDs (of the form `<driver name>/<pool name>/<device name>`) along with some extra metadata indicating its health status. Just as before, a device completely disappearing would still need to trigger some state change, but now more informative information could be attached in the form of metadata when a device isn't necessarily gone, but also isn't operating as it should be.

Reply:
> Does it makes sense to keep some sort of stream like this, but make it just about health status information? 

Yes. It can be in an optional gRPC interface, so DRA drivers can opt into this by implementing it.

I would limit it to "prepared" devices and include the claim name/namespace/UID. That should be enough information for kubelet to correlate with the pods for which the claim was prepared and then post that information for those pods.

Reply:
So a separate interface, not tied to the one that defined `NodePrepareResources()` and `NodeUnprepareResources()`?

Reply:
Added